packages/openapi3/src/schema-emitter-3-0.ts

```diff
@@ -0,0 +1,266 @@
+    }
+
+    const maxValueExclusive = getMaxValueExclusive(program, type);
+    if (maxValueExclusive !== undefined) {
+      target.maximum = maxValueExclusive;
+      target.exclusiveMaximum = true;
+    }
+
+    this.#applySchemaExamples(type, target);
+  }
+
+  enumSchema(en: Enum): OpenAPI3Schema {
+    const program = this.emitter.getProgram();
+    if (en.members.size === 0) {
+      reportDiagnostic(program, { code: "empty-enum", target: en });
+      return {};
+    }
+
+    const enumTypes = new Set<JsonType>();
+    const enumValues = new Set<string | number>();
+    for (const member of en.members.values()) {
+      enumTypes.add(typeof member.value === "number" ? "number" : "string");
+      enumValues.add(member.value ?? member.name);
+    }
+
+    if (enumTypes.size > 1) {
+      reportDiagnostic(program, { code: "enum-unique-type", target: en });
+    }
+
+    const schema: OpenAPI3Schema = {
+      type: enumTypes.values().next().value!,
+      enum: [...enumValues],
+    };
+
+    return this.applyConstraints(en, schema);
+  }
+
+  unionSchema(union: Union): ObjectBuilder<OpenAPI3Schema> {
+    const program = this.emitter.getProgram();
+    if (union.variants.size === 0) {
+      reportDiagnostic(program, { code: "empty-union", target: union });
+      return new ObjectBuilder({});
+    }
+    const variants = Array.from(union.variants.values());
+    const literalVariantEnumByType: Record<string, any[]> = {};
+    const ofType = getOneOf(program, union) ? "oneOf" : "anyOf";
+    const schemaMembers: { schema: any; type: Type | null }[] = [];
+    let nullable = false;
+    const discriminator = getDiscriminator(program, union);
+    const isMultipart = this.getContentType().startsWith("multipart/");
+
+    for (const variant of variants) {
+      if (isNullType(variant.type)) {
+        nullable = true;
+        continue;
+      }
+
+      if (isMultipart && isBytesKeptRaw(program, variant.type)) {
+        schemaMembers.push({ schema: { type: "string", format: "binary" }, type: variant.type });
+        continue;
+      }
+
+      if (isLiteralType(variant.type)) {
+        if (!literalVariantEnumByType[variant.type.kind]) {
+          const enumValue: any[] = [variant.type.value];
+          literalVariantEnumByType[variant.type.kind] = enumValue;
+          schemaMembers.push({
+            schema: { type: literalType(variant.type), enum: enumValue },
+            type: null,
+          });
+        } else {
+          literalVariantEnumByType[variant.type.kind].push(variant.type.value);
+        }
+      } else {
+        const enumSchema = this.emitter.emitTypeReference(variant.type, {
+          referenceContext: isMultipart ? { contentType: "application/json" } : {},
+        });
+        compilerAssert(enumSchema.kind === "code", "Unexpected enum schema. Should be kind: code");
+        schemaMembers.push({ schema: enumSchema.value, type: variant.type });
+      }
+    }
+
+    const wrapWithObjectBuilder = (
```

Comment:
I think this section will actually end up changing for 3.1 since 3.1 allows properties set as siblings to a `$ref` to override the ref's properties. Also don't need the `nullable` piece for 3.1. I didn't want to spend too much time refactoring this to share the code given I expect these to diverge in the short term - but would be happy to take another look before this goes into `main` if they end up not diverging.